File: Biblioteca-main/Backend/models/Libro.js
const mongoose = require('mongoose');

const libroSchema = new mongoose.Schema({
  titulo: {
    type: String,
    required: [true, 'El título es obligatorio'],
    trim: true
  },
  autor: {
    type: String,
    required: [true, 'El autor es obligatorio'],
    trim: true
  },
  genero: {
    type: String,
    required: true,
    trim: true
  },
  descripcion: {
    type: String,
    required: true
  },
  anio_edicion: {
    type: String,
    required: true
  },
  imagen: {
    type: String,
    default: 'https://via.placeholder.com/150'
  },
  disponibilidad: {
    type: String,
    enum: ['Disponible', 'Agotado', 'Prestado'],
    default: 'Disponible'
  },
  copias_totales: {
    type: Number,
    default: 1,
    min: 0
  },
  copias_disponibles: {
    type: Number,
    default: 1,
    min: 0
  }
}, {
  timestamps: true
});

function syncDisponibilidad(libro) {
  if (typeof libro.copias_totales === 'number' && typeof libro.copias_disponibles === 'number') {
    if (libro.copias_disponibles > libro.copias_totales) {
      libro.copias_disponibles = libro.copias_totales;
    }
  }

  libro.disponibilidad = (libro.copias_disponibles ?? 0) > 0 ? 'Disponible' : 'Agotado';
}

libroSchema.pre('save', function(next) {
  syncDisponibilidad(this);
  next();
});

libroSchema.methods.actualizarDisponibilidad = function() {
  syncDisponibilidad(this);
};

module.exports = mongoose.model('Libro', libroSchema);

File: Biblioteca-main/Movil/biblioteca_flutter/lib/models/libro.dart
class Libro {
  final String id;
  final String titulo;
  final String autor;
  final String genero;
  final String disponibilidad;
  final int copiasTotales;
  final int copiasDisponibles;
  final String? imagen;
  final String? descripcion;
  final String? anioEdicion;

  Libro({
    required this.id,
    required this.titulo,
    required this.autor,
    required this.genero,
    required this.disponibilidad,
    required this.copiasTotales,
    required this.copiasDisponibles,
    this.imagen,
    this.descripcion,
    this.anioEdicion,
  });

  bool get estaDisponible =>
      disponibilidad.toLowerCase().contains('disponible') && copiasDisponibles > 0;

  factory Libro.fromJson(Map<String, dynamic> json) {
    return Libro(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      titulo: json['titulo'] ?? '',
      autor: json['autor'] ?? '',
      genero: json['genero'] ?? '',
      disponibilidad: json['disponibilidad'] ?? 'Desconocido',
      copiasTotales: json['copias_totales'] is int
          ? json['copias_totales']
          : int.tryParse('${json['copias_totales'] ?? 0}') ?? 0,
      copiasDisponibles: json['copias_disponibles'] is int
          ? json['copias_disponibles']
          : int.tryParse('${json['copias_disponibles'] ?? 0}') ?? 0,
      imagen: json['imagen'],
      descripcion: json['descripcion'],
      anioEdicion: json['anio_edicion'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'titulo': titulo,
      'autor': autor,
      'genero': genero,
      'disponibilidad': disponibilidad,
      'copias_totales': copiasTotales,
      'copias_disponibles': copiasDisponibles,
      'imagen': imagen,
      'descripcion': descripcion,
      'anio_edicion': anioEdicion,
    };
  }
}
